tests/foreman/cli/test_remoteexecution.py

```diff
@@ -65,7 +62,7 @@ def fixture_org():
 
 
 @pytest.fixture(params=distros, scope="module")
-def fixture_vmsetup(request, fixture_org):
+def fixture_vmsetup(request, fixture_org, subnet_id=None, by_ip=True):
```

Comment:
How are `subnet_id` and `by_ip` supposed to be set?

As far as I know, fixture has special properties that separate it from normal python function. Fixture is always pulled in by test as a dependency, and as such, is always referenced unparametrized. Fixture can be parametrized, but then all combinations of all parameters all called, creating multiple fixtures in process, and each test depending on that fixture is called multiple times (once for every unique combination).

To allow test to set `subnet_id` and `by_ip` (which I assume is what you want), you would need to modify this fixture to return function creating VM with parameters. Then tests would need to call that function explicitly:
```python
@pytest.fixture
def fixture_vmsetup(request):
    def inner(subnet_id=None, by_ip=True):
        # do what you have to do
        return created_vm
    return inner  # return reference to function - nothing is called yet at this point

def test_whatever(fixture_vmsetup):
    vm = fixture_vmsetup()  # call function returned by fixture
    # OR vm = fixture_vmsetup(by_ip=False)
```
One drawback of that approach is that every single test referencing this fixture would need to be modified, i.e. instead of `self.client = fixture_vmsetup` now you need `self.client = fixture_vmsetup()` (as fixture returns function, and you want to call that function and assign returned value.

Another problem is that in naive implementation (above), vm will be created independently in each function-returned-by-fixture invocation. So you need to introduce storage for created vms. I would go with "one VM for each unique combination of parameters" (so all tests calling `fixture_vmsetup()` will get the same VM). Of course that mandates re-implementation of teardown mechanism, as you are no longer using yield.

Reply:
Hi @mirekdlugosz, these are good points but I wasn't really aiming for per-test parametrization, rather a way to easily turn off by_ip for the class when needed (as it will probably out or support under IPv6). I don't expect it will be worth to test by_ip along with by_hostname, I was thinking that the subnet_id and by_ip settings would be set globally according to some settings in robottelo.properties.

Reply:
@pondrejk I agree with @mirekdlugosz  approach on it.
However another side of coin , afaik all rex tasks and tests in automation are running(get success) only by setting `remote_execution_connect_by_ip True` at host level hence as per my understanding we should to go with below straight forward approach and if someone really need to set this param to `False` at test level than this parameter could be override at test level it self ..
i.e.
```
@pytest.fixture(params=distros, scope="module")
def fixture_vmsetup(request, fixture_org):
    """Create Org, Lifecycle Environment, Content View, Activation key,
    VM, install katello-ca, register it, add remote execution key
    """
    # Create VM and register content host
    client = VirtualMachine(distro=request.param)
    try:
        client.create()
        if request.param in [DISTRO_SLES11, DISTRO_SLES12]:
            # SLES hostname in subscription-manager facts doesn't include domain
            client._hostname = client.hostname.split(".")[0]
        client.install_katello_ca()
        # Register content host
        client.register_contenthost(
            org=fixture_org.label,
            lce='Library'
        )
        assert client.subscribed
        add_remote_execution_ssh_key(client.ip_addr)
        Host.update({
            'name': client.hostname,
            'subnet-id': subnet_id,
            })
        Host.set_parameter({
            'host': client.hostname,
            'name': 'remote_execution_connect_by_ip',
            'value': 'True',
            })
        yield client
    finally:
        client._hostname = None
        client.destroy()
```

Reply:
@mirekdlugosz @vijay8451 how about making the by_ip setting a fixture parameter before fixture_vmsetup like it is currently done with distros? This will allow 3 scenarios: run all tests with by_ip, without by_ip or with both.

Reply:
@pondrejk could you please share an example how exactly you wants to achieve this  ... as that would help to understand the idea around it.

Reply:
@pondrejk any luck with this?

Reply:
@mirekdlugosz @vijay8451 @JacobCallahan looking back at this, I'd rework this to be put in after branching. I'd keep the PR open to preserve the discussion if that's OK.